README.md
# ![Status: Entwurf](https://img.shields.io/badge/Status-Entwurf-yellow.svg) SMITH VHF Vorhofflimmern

## Einführung

Dieses Projekt führt das SMITH DUP Vorhofflimmern (VHF) aus. Dafür erzeugt dieses Skript zwei Tabellen mit den für die
Analyse benötigten Inhalten. Diese Tabellen sollen entweder zentral erzeugt und an die datenauswertendende Stelle
übergeben oder dezentral in den DIZen analysiert werden, das Analyseergebnis muss in diesem Fall manuell an die
Biometrie übergeben werden.

Das Readme beschreibt zunächst die technischen Details der Verwendung. Darunter sind die verwendeten
CodeSysteme/Ressourcen/Profile und der konzeptionelle Ablauf der Abfrage beschrieben.

## Verwendung

### dupctl (Docker)

Eine einfache und reproduzierbare Ausführung der SMITH PheP DUPs wird über das dupctl Command Line Interface (cli)
sichergestellt. Nach der Installation der cli (siehe [dupctl][dupctl]) wird das DUP Skript über eine Kommandozeile
wie folgt gestartet:

```bash
dupctl retrieve --dup vhf [...]
```

### Manuelle Ausführung

Bei Bedarf können DUPs auch ohne den Einsatz von Docker durchgeführt werden.

*Die manuelle Ausführung ist **nicht** durch das downloaden/klonen dieses Repositories möglich!*
Stattdessen muss das Workpackage in Form eines Archivs mittels folgenden Links heruntergeladen werden. Das Archiv
enthält alle nötigen Skripte und Libraries für eine manuelle Ausführung mittels R.

<div align="center"><a href="https://git.smith.care/smith/uc-phep/dup/vhf/-/jobs/artifacts/master/download?job=build-archive:latest">Download als ZIP Archiv</a></div>

* **R Version: 4.2.0**

* **CRAN Snaphot Datum: 2022-06-22**

Die weiteren Schritte der manuellen Ausführung sind in der [DUP README][readme] erläutert.

## Output

Das Skript erzeugt mehrere Ordner im Projekt-Directory. Um für den Projectathon eine möglichst einfache übersichtliche
Lösung zu bekommen, werden alle files, die darin erzeugt werden bei mehrmaligem Ausführen ggf. einfach überschrieben.

### Ergebnisse

Wenn die Abfrage erfolgreich durchgeführt wurde, sind hier zwei semikolongetrennte csv-Dateien (= lässt sich durch
Doppelklick in Excel öffnen) zu finden. Die enthaltenen Variablen sind hier kurz erklärt:

**Kohorte.csv**

Diese Tabelle enthält eine Kombination von Informationen aus der Patient Ressource, der Encounter Ressource und der
Observation Ressource. Sie enthält alle Fälle, für die es eine Observation mit einer NTproBNP Messung im geforderten
Zeitraum gibt.

| Variable                             | Bedeutung                                                                                                                                           |
|--------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------|
| subject                              | Logical id der Patient Ressource                                                                                                                    |
| encounter.start                      | Startzeitpunkt des Einrichtungskontakt-Encounters, der zeitlich zur NTproBNP-Messung gehört.                                                        |
| encounter.end                        | Stoppzeitpunkt des Einrichtungskontakt-Encounters, der zeitlich zur NTproBNP-Messung gehört.                                                        |
| serviceType                          | ServiceType (Stationsschlüssel) des Einrichtungskontakt-Encounters, der zeitlich zur NTproBNP-Messung gehört.                                       |
| NTproBNP.date                        | Datum (effectiveDateTime) der Observation mit der NTproBNP-Messung.                                                                                 |
| NTproBNP.valueQuantity.value         | Numerischer Wert (valueQuantity.value) der Observation mit der NTproBNP-Messung.                                                                    |
| NTproBNP.valueQuantity.comparator    | Komparator (valueQuantity.comparator, falls vorhanden) der Observation mit der NTproBNP-Messung, welcher den numerischen Wert qualifiziert.         |
| NTproBNP.valueCodeableConcept.code   | Kodierter Wert (valueCodeableConcept.code) der Observation mit der NTproBNP-Messung, für Fälle in denen die Messung nicht numerisch abgelegt wurde. |
| NTproBNP.valueCodeableConcept.system | Codesystem zu NTproBNP.valueCodeableConcept.code                                                                                                    |
| NTproBNP.code                        | Loinc-Code (code.coding.code) der Observation mit der NTproBNP-Messung.                                                                             |
| NTproBNP.codeSystem                  | CodeSystem (code.coding.system) der Observation mit der NTproBNP-Messung.                                                                           |
| NTproBNP.unit                        | Code der Einheit (valueQuantity.code) der NTproBNP-Messung.                                                                                         |
| NTproBNP.unitSystem                  | Codesystem der Einheit (valueQuantity.code) der NTproBNP-Messung.                                                                                   |
| gender                               | Geschlecht (gender) der Patient Ressource                                                                                                           |
| birthdate                            | Geburtsdatum (birthDate) der Patient Ressource                                                                                                      |

**Diagnosen.csv**

Diese Tabelle enthält alle Diagnosen der Patienten aus Kohorte.csv, die einen der im Antrag genannten ICD-Kodes
enthalten.

| Variable                  | Bedeutung                                                                                            |
|---------------------------|------------------------------------------------------------------------------------------------------|
| condition.id              | Ressourcen ID der Condition Ressource                                                                |
| clinicalStatus.code       | Code des ClinicalStatus                                                                              |
| clinicalStatus.system     | Codesystem des ClinicalStatus                                                                        |
| verificationStatus.code   | Code des verificationStatus                                                                          |
| verificationStatus.system | CodeSystem des verificationStatus                                                                    |
| code                      | ICD-Code                                                                                             |
| code.system               | Codesystem des ICD Codes                                                                             |
| subject                   | ID der zugehörigen Patient Ressource                                                                 |
| encounter.id              | ID der zugehörigen Encounter Ressource                                                               |
| diagnosis.use.code        | Encounter.diagnosis.use.coding.code mit dem die Condition im zugehörigen Encounter beschrieben ist   |
| diagnosis.use.system      | Encounter.diagnosis.use.coding.system mit dem die Condition im zugehörigen Encounter beschrieben ist |

**smith_select.log**

Neben den Ergebnistabellen wird außerdem eine "smith_select.log"-Datei erzeugt, welche die Anzahl der extrahierten
Fälle, Patienten und die Laufzeit des R-Skriptes dokumentiert. Das log-file muss nicht geteilt werden, es dient den
DIZen nur als Hilfestellung für die Einschätzung von Laufzeiten und Ergebnismengen.

### Bundles

Dieser Ordner enthält die heruntergeladenen Bundles und kann der Kontrolle dienen, falls die Tabellen nicht aussehen wie
erwartet.

### errors

Dieser Ordner enthält ggf. Fehlermeldungen, wenn die Abfrage nicht erfolgreich durchgeführt werden kann.

## Verwendete Codesysteme

Das Skript verwendet folgende Codesysteme:

- *http://loinc.org* für `Observation.code.coding.system` -> Dieses System wird für den Download per FHIR Search
  verwendet
- *urn:oid:2.16.840.1.113883.3.1937.777.24.5.3* für `Consent.provision.provision.code.coding.system` -> Nur verwendet,
  wenn konsentierte Daten über Consent Ressource selektiert werden, also wenn `filterConsent <- TRUE` in config.R.

## Verwendete Profile/Datenelemente

Die Abfragen sind auf Basis der MII Profile für die entsprechenden Ressouren geschrieben. Die Skripte sind kompatibel
mit dem jeweils neuesten Release der verfügbaren Major Versionen (1 und, wenn vorhanden, 2). In den meisten Fällen
werden die Skripte auch funktionieren, wenn nicht genau dieser Release des Profils verwendet wird. Es ist jedoch
zwingend notwendig, dass die Ressourcen das jeweilige Profil im `Resource.meta.profile` Element benennen. Im Folgenden
ist für jeden verwendeten Ressourcentyp beschrieben, welche Elemente für die FHIR-Search-Abfrage an den Server verwendet
werden (diese Elemente *müssen* vorhanden sein, damit kein Fehler geworfen wird) und welche Elemente im Skript
extrahiert und in die Ergebnistabellen geschrieben werden.

### Modul Labor: Observation

Profil: `https://www.medizininformatik-initiative.de/fhir/core/modul-labor/StructureDefinition/ObservationLab`

Version: [1.0.6](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.laborbefund/1.0.6/~introduction)

Für Serverabfrage verwendete Elemente:

- `Observation.subject`
- `Observation.code`
- `Observation.effective`
- `Observation.meta.profile`

Extrahierte Elemente:

- `Observation.effectiveDateTime`
- `Observation.subject.reference`
- `Observation.code.coding.code`
- `Observation.code.coding.system`
- `Observation.valueQuantity.value`
- `Observation.valueQuantity.code`
- `Observation.valueQuantity.system`

### Modul Person: Patient

Profil: `https://www.medizininformatik-initiative.de/fhir/core/modul-person/StructureDefinition/Patient`

Version: [2.0.0-alpha3](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.person/2.0.0-alpha3)
bzw. [1.0.14](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.person/1.0.14)

Für Serverabfrage verwendete Elemente:

- keine

Extrahierte Elemente:

- `Patient.id`
- `Patient.gender`
- `Patient.birthDate`

### Modul Fall: Encounter

Profil: `https://www.medizininformatik-initiative.de/fhir/core/modul-fall/StructureDefinition/KontaktGesundheitseinrichtung`

Version: [1.0.1](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.fall/1.0.1)

Für Serverabfrage verwendete Elemente:

- `Encounter.meta.profile`
- `Encounter.subject.reference`

Extrahierte Elemente:

- `Encounter.subject.reference`
- `Encounter.period.start `
- `Encounter.period.end`
- `Encounter.serviceType.coding.display`
- `Encounter.diagnosis.use.coding.code`
- `Encounter.diagnosis.use.coding.system`

### Modul Diagnose: Condition

Profil: `https://www.medizininformatik-initiative.de/fhir/core/modul-diagnose/StructureDefinition/Diagnose`

Version: [2.0.0-alpha3](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.diagnose/2.0.0-alpha3)
bzw. [1.0.4](https://simplifier.net/packages/de.medizininformatikinitiative.kerndatensatz.diagnose/1.0.4)

Für Serverabfrage verwendete Elemente:

- `Condition.meta.profile`
- `Condition.subject.reference`

Extrahierte Elemente:

- `Condition.clinicalStatus.coding.code`
- `Condition.clinicalStatus.coding.system`
- `Condition.verificationStatus.coding.code`
- `Condition.verificationStatus.coding.system`
- `Condition.code.coding.code`
- `Condition.code.coding.system`
- `Condition.subject.reference`
- `Conditions.encounter.reference`

### Modul Consent: Consent

Wird nur verwendet, wenn `filterConsent <- TRUE` in `config.R`, d.h. wenn konsentierte und nicht konsentierte Daten
durch das R-Skript gefiltert werden müssen und die Trennung nicht durch unterschiedliche FHIR-Repositories erfolgt.

Profil: `http://fhir.de/ConsentManagement/StructureDefinition/Consent`
Version: [1.0.0](https://simplifier.net/packages/de.einwilligungsmanagement/1.0.0)
Beispiel: https://simplifier.net/packages/de.einwilligungsmanagement/1.0.0/files/405292

Für Serverabfrage verwendete Elemente:

- `Consent.patient.reference`

Extrahierte Elemente:

- `Consent.patient.reference`
- `Consent.provision.provision.code.coding.code`
- `Consent.provision.provision.code.coding.system`

## Konzeptioneller Ablauf der Abfrage

Prinzipiell geht das Skript wie folgt vor:

1) Lade alle Observations, die eine NTproBNP-Messung darstellen (`loinc code` ist einer
   aus: `33763-4,71425-3,33762-6,83107-3, 83108-1,77622-9,77621-1`), welche ein `effective` im
   Zeitraum `2019-01-01 - 2021-12-31` haben und das
   Profil `https://www.medizininformatik-initiative.de/fhir/core/modul-labor/StructureDefinition/ObservationLab`
   erfüllen, herunter. Ziehe dazu außerdem (mit `_include`) die zugehörigen Patient Ressourcen.

- `[base]/Observation?_include=Observation:patient&_profile=https://www.medizininformatik-initiative.de/fhir/core/modul-labor/StructureDefinition/ObservationLab&code=http://loinc.org|33763-4,http://loinc.org|71425-3,http://loinc.org|33762-6,http://loinc.org|83107-3,http://loinc.org|83108-1,http://loinc.org|77622-9,http://loinc.org|77621-1&date=ge2019-01-01&date=le2021-12-31`

2) Optional: Lade die zu den Patienten gehörigen Consents herunter und filtere die bisher geladenen Daten, sodass nur
   Daten von Patienten mit Consent übrig bleiben.

- `[base]/Consent?patient=xxx%2Cyyy`

3) Extrahiere die IDs der Patient Ressourcen und verwende sie um alle zugehörigen Encounter und Condition-Ressourcen
   herunterzuladen. Dafür werden die Abfragen soweit gesplittet, dass der jeweilige GET-Request eine Länge von 1800
   Zeichen nicht überschreitet.(Mir ist klar, das POST eleganter ist, aber wir versuchen die Anforderungen an den
   Server/Rechte möglichst niedrig zu halten.)

- `[base]/Encounter?&_profile=https://www.medizininformatik-initiative.de/fhir/core/modul-fall/StructureDefinition/KontaktGesundheitseinrichtung&subject=xxx,yyy`
- `[base]/Condition?_profile=https://www.medizininformatik-initiative.de/fhir/core/modul-diagnose/StructureDefinition/Diagnose&subject=xxx,yyy`


4) Filtere die Encounter, sodass nur die Encounter übrig bleiben, in deren `period` eine NTproBNP Observation liegt.

5) Filtere die Diagnosen, sodass nur Diagnosen übrig bleiben, die zu den Encountern aus 4) gehören.


[dupctl]: https://git.smith.care/smith/uc-phep/dup-control
[readme]: https://git.smith.care/smith/uc-phep/dup/readme


CHANGELOG.md
# Changelog

*Note: This changelog reflects the changes until first released projectathon version*

**28.03.2022**

*Änderung*: Zusätzlich zu den Ergebnis-Tabellen wird nun ein Textfile "Ergebnisse/smith_select.log" erzeugt, welches die
Anzahl der extrahierten Fälle, Patienten und die Laufzeit des R-Skriptes dokumentiert. Das log-file muss nicht geteilt
werden, es dient den DIZen nur als Hilfestellung für die Einschätzung von Laufzeiten und Ergebnismengen.

----------------------------

**17.03.2022**

*Änderung*: In der Variable `serviceType` wird jetzt das Element `Encounter.serviceType.coding.display` extrahiert,
anstatt wie vorher `Encounter.serviceType`, was nicht zu einem extrahierbaren einzelnen String geführt hätte. Die Form
und der inhalt der Tabelle ändert sich dadurch nicht, nur der Inhalt der betreffenden Variable, die jetzt das enthält,
was intendiert/kommuniziert war, statt eines `NA`.

----------------------------

**14.03.2022**

*Änderung*: Es werden nicht mehr nur numerische NTproBNP-Messwerte aus dem Element `Observation.valueQuantity.value`
extrahiert, sondern zusätzlich auch die Elemente `Observation.valueQuantity.comparator`
, `Observation.valueCodeableConcept.coding.code` und `Observation.valueCodeableConcept.coding.system`, um auch Messwerte
abzudecken, die sich nicht in `valueQuantity.value` allein abbilden lassen, z.b. Angaben wie `<50`.
Die Spalten, die diese Informationen insgesamt abdecken heißen `NTproBNP.valueQuantity.value`
, `NTproBNP.valueQuantity.comparator`, `NTproBNP.valueCodeableConcept.code`, `NTproBNP.valueCodeableConcept.system`.

----------------------------

**04.03.2022**

*Änderung*: Zur Qualitätssicherung wird für jede NTproBNP-Messung nun auch der zugehörige Loinc-Code extrahiert, der zur
Filterung der jeweiligen Observation verwendet wurde. Es gibt deshalb in Kohorte.csv nun zwei zusätzliche Spalten:
NTproBNP.code und NTproBNP.codeSystem.

----------------------------

**18.02.2022**

*Änderung*: Das Skript schickt jetzt einige informative Nachrichten in die Konsole um das Debugging zu erleichtern. Die
Ergebnisse ändern sich dadurch in keiner Weise.

----------------------------

**19.01.2022**

*Änderung*: Logik beim herunterladen von Conditions geändert: Es werden jetzt alle Conditions zu den untersuchten
Patienten gezogen und anschließend so gefiltert, dass nur Conditions übrig bleiben, die zu den gewünschten Encountern
gehören. Entsprechend wurde im `config.R` das Profil für die Condition ergänzt und kann nach Bedarf an- und abgeschaltet
werden.

*Erklärung*: Damit ist es jetzt irrelevant, ob der Encounter auf die Condition verlinkt oder die Condition auf den
Encounter verlinkt. Das Skript funktioniert, solange mindestens eine der Richtungen gegeben ist. Diese Änderung wurde
implementiert, weil sich herausstellt, dass die Linkrichtung in den verschiedenen DIZen zu heterogen ist, als das man
sich auf eine von beiden verlassen könnte.

----------------------------

**14.01.2022**

*Änderung*: Typo in `smith_select.R` korrigiert.

*Erklärung* Korrigiert Fehler in der zeitlichen Relation im merge-Befehl in Z301. Sollte keine oder maximal
geringfügigen Einfluss auf die Ergebnisse haben, da im Folgenden sowieso nochmal nach korrektem Zeitbezug gefiltert
wird.

----------------------------

**13.12.21**

*Änderung*: Changelog im README wurde ergänzt

-------------------

**10.12.21**

*Änderung*: Nicht mehr benötigte Zwischenergebnisse werden frühzeitig gelöscht.

*Erklärung*: In DIZen mit großen Datenmengen kommt der Arbeitsspeicher des Docker-Containers an seine Grenzen. Die
Änderung hat keinen Einfluss auf die Ergebnisse der Skripte.

---------------------

**07.12.21**

*Änderung*: Encounter-Ressourcen, die durch `_include` doppelt heruntergeladen wurden, werden jetzt im Skript gelöscht.

*Erklärung*: Doppelt heruntergeladene Encounter führten zuvor im Skript für Fehlermeldungen beim Mergen der Daten.
DIZen, die das Skript zuvor schon fehlerfrei ausführen konnten, hatten keine doppelt heruntergeladenen Encounter und
bekommen durch diese Änderung demzufolge auch keine geänderten Ergebnisse

---------------------------

*Änderung*: Merge Encounter- und Observation-Ressouren basierend auf Subject-ID und Datum, nicht auf Subject-ID allein.

*Erklärung*: Der Merge anhand der Subject-ID allein funktionierte nur eindeutig, solange Encounter und Observations
eines Subject (=Patient) in einem 1:n oder m:1-Verhältnis standen. In diesem Fall konnte man erst mergen und dann die
Zeiträume filtern, sodass nur zueinander passende Ressourcen behalten wurden. Wenn Encounter und Observations in m:n
Verhältnis vorlagen, warf das Sktipt einen Fehler, der durch die neue Merge-Technik behoben wird.
DIZen, bei denen das Sktipt schon vor der Anpassung keinen Fehler warf, bekommen trotzdem das gleiche Ergebnis, weil die
zeitliche Filterung der Ressourcen jetzt einfach nur einen Arbeitsschritt nach vorne gelegt wurde.

----------------------------

*Änderung*: Die Einschränkung der Abfrage auf MII-Profile z.B.
über `_profile=https://www.medizininformatik-initiative.de/fhir/core/modul-fall/StructureDefinition/KontaktGesundheitseinrichtung`
ist jetzt im `config.R` konfigurierbar. Sie kann komplett ausgeschaltet oder wenn nötig auf ein andere Profil angepasst
werden.

*Erklärung*: Einige DIZen haben die Profile noch nicht in `Resource.meta.profile` referenziert und bekommen mit der
Einschränkung keine Ergebnisse.

-------------------------------


smith_select.R
### Preparation
start <- Sys.time()
# load/install a packages
source("install_R_packages.R")

#create directories
if (!dir.exists("Ergebnisse")) {
  dir.create("Ergebnisse")
}
if (!dir.exists("errors")) {
  dir.create("errors")
}
if (!dir.exists("Bundles")) {
  dir.create("Bundles")
}


# source config
if (file.exists("config.R") && !dir.exists("config.R")) {
  source("config.R")
} else{
  source("config.R.default")
}

#If needed disable peer verification
if (!ssl_verify_peer) {
  httr::set_config(httr::config(ssl_verifypeer = 0L))
}

#remove trailing slashes from endpoint
base <-
  if (grepl("/$", base)) {
    strtrim(base, width = nchar(base) - 1)
  } else{
    base
  }

# Brackets around indexes for nested values after fhir_crack()
brackets = c("[", "]")
sep = " || "

### Get all Observations between 2019-01-01 and 2021-12-31 with loinc 33763-4,71425-3,33762-6,83107-3, 83108-1, 77622-9,77621-1
# also get associated patient resources --> initial patient population
# Observations have to implement MII profile
# TODO: weitere LOINC Codes ergänzen zB für proBNP
obs_request <- fhir_url(
  url = base,
  resource = "Observation",
  parameters = c(
    "code" = "http://loinc.org|33763-4,http://loinc.org|71425-3,http://loinc.org|33762-6,http://loinc.org|83107-3,http://loinc.org|83108-1,http://loinc.org|77622-9,http://loinc.org|77621-1",
    "date" = "ge2019-01-01",
    "date" = "le2021-12-31",
    "_include" = "Observation:patient"
  )
)

#add profile from config
obs_request <- fhir_url(paste0(obs_request, obs_profile))

# download bundles
message("Downloading Observations: ", obs_request, "\n")
obs_bundles <- fhir_search(
  request = obs_request,
  username = username,
  password = password,
  token = token,
  log_errors = "errors/observation_error.xml",
  verbose = 0,
  max_bundles = 3
)

#save for checking purposes
#TODO löschen?
fhir_save(bundles = obs_bundles, directory = "Bundles/Observations")

# flatten
obs_description <- fhir_table_description(
  "Observation",
  cols = c(
    NTproBNP.date = "effectiveDateTime",
    subject = "subject/reference",
    NTproBNP.valueQuantity.value = "valueQuantity/value",
    NTproBNP.valueQuantity.comparator = "valueQuantity/comparator",
    NTproBNP.valueCodeableConcept.code = "valueCodeableConcept/coding/code",
    NTproBNP.valueCodeableConcept.system = "valueCodeableConcept/coding/system",
    NTproBNP.code = "code/coding/code",
    NTproBNP.codeSystem = "code/coding/system",
    NTproBNP.unit = "valueQuantity/code",
    NTproBNP.unitSystem = "valueQuantity/system"
  )
)

pat_description <- fhir_table_description("Patient",
                                          cols = c(
                                            id = "id",
                                            gender = "gender",
                                            birthdate = "birthDate"
                                          ))

message("Cracking ", length(obs_bundles), " Observation Bundles.\n")
obs_tables <- fhir_crack(
  obs_bundles,
  design = fhir_design(obs = obs_description, pat = pat_description),
  sep = sep,
  brackets = brackets,
  data.table = TRUE,
  verbose = 0
)

rm(obs_bundles)

if (nrow(obs_tables$obs) == 0) {
  write(
    "Konnte keine Observations für NTproBNP auf dem Server finden. Abfrage abgebrochen.",
    file = "errors/error_message.txt"
  )
  stop("No NTproBNP Observations found - aborting.")
}

if (nrow(obs_tables$pat) == 0) {
  write(
    "Konnte keine Patientenressourcen für NTproBNP-Observations auf dem Server finden. Abfrage abgebrochen.",
    file = "errors/error_message.txt"
  )
  stop("No Patients for NTproBNP Observations found - aborting.")
}

# melt multiple entries/remove indices
obs_tables$pat <-
  fhir_rm_indices(obs_tables$pat, brackets = brackets)

obs_tables$obs <- fhir_melt(
  obs_tables$obs,
  columns = c("NTproBNP.code", "NTproBNP.codeSystem"),
  brackets = brackets,
  sep = sep,
  all_columns = TRUE
)

obs_tables$obs <- fhir_melt(
  obs_tables$obs,
  columns = c("NTproBNP.code", "NTproBNP.codeSystem"),
  brackets = brackets,
  sep = sep,
  all_columns = TRUE
)

obs_tables$obs <-
  fhir_rm_indices(obs_tables$obs, brackets = brackets)

obs_tables$obs[, resource_identifier := NULL]

obs_tables$obs <-
  obs_tables$obs[NTproBNP.codeSystem == "http://loinc.org"]

# get rid of resources that have been downloaded multiple times via _include
obs_tables$pat <- unique(obs_tables$pat)

### Prepare Patient id from initial patient population for Search requests that download associated resources (e.g. consent, encounters, conditions)

### merge observation and patient data
# prepare key variables for merge
obs_tables$obs[, subject := sub("Patient/", "", subject)]

# sort out col types
obs_tables$obs[, NTproBNP.date := as.Date(NTproBNP.date)]

# merge
message(
  "Merging Observation and Patient data based on Patient id.\n",
  "Number of unique Patient ids in Patient data: ",
  length(unique(obs_tables$pat$id)),
  " in ",
  nrow(obs_tables$pat),
  " rows",
  "\n",
  "Number of unique Patient ids in Observation data: ",
  length(unique(obs_tables$obs$subject)),
  " in ",
  nrow(obs_tables$obs),
  " rows",
  "\n"
)

obsdata <- merge.data.table(
  x = obs_tables$obs,
  y = obs_tables$pat,
  by.x = "subject",
  by.y = "id",
  all.x = TRUE
)

rm(obs_tables)

#split patient id list into smaller chunks that can be used in a GET url
#(split because we don't want to exceed allowed URL length)
#TODO magic Number nach oben
patients <- obsdata$subject #all patient ids

nchar_for_ids <- 1800 - nchar(
  paste0(
    base,
    "Encounter?_profile=https://www.medizininformatik-initiative.de/fhir/core/modul-fall/StructureDefinition/KontaktGesundheitseinrichtung"
  )
) #assume maximal length of 1800

n <- length(patients)

list <-
  split(patients, ceiling(seq_along(patients) / n)) #split patients ids in chunks of size n

nchar <-
  sapply(list, function(x) {
    sum(nchar(x)) + (length(x) - 1)
  }) #compute number of characters for each chunk, including commas for seperation

#reduce the chunk size until number of characters is small enough
while (any(nchar > nchar_for_ids)) {
  n <- n / 2
  list <- split(patients, ceiling(seq_along(patients) / n))
  nchar <- sapply(list, function(x) {
    sum(nchar(x)) + (length(x) - 1)
  })
}


# get encounters and diagnoses
# --> all encounters and diagnoses of initial patient population,
# has be filtered to only include encounters with NTproBNP Observation later on
encounter_bundles <- list()
condition_bundles <- list()

message("Downloading Encounters and Conditions.\n")

invisible({
  lapply(list, function(x) {
    #x<-list[[1]]
    ids <- paste(x, collapse = ",")
    
    ### Encounters
    enc_request <- fhir_url(
      url = base,
      resource = "Encounter",
      parameters = c(subject = ids,
                     type = "einrichtungskontakt")
    )
    
    #add profile from config
    enc_request <- fhir_url(url = paste0(enc_request, enc_profile))
    
    
    encounter_bundles <<- append(
      encounter_bundles,
      fhir_search(
        enc_request,
        username = username,
        password = password,
        token = token,
        log_errors = "errors/encounter_error.xml",
        verbose = 0
      )
    )
    
    ### Conditions
    con_request <- fhir_url(
      url = base,
      resource = "Condition",
      parameters = c(subject = ids)
    )
    
    #add profile from config
    con_request <- fhir_url(url = paste0(con_request, con_profile))
    
    
    condition_bundles <<- append(
      condition_bundles,
      fhir_search(
        con_request,
        username = username,
        password = password,
        token = token,
        log_errors = "errors/condition_error.xml",
        verbose = 0
      )
    )
    
  })
})

# bring encounter results together, save and flatten
encounter_bundles <-
  fhircrackr:::fhir_bundle_list(encounter_bundles)

condition_bundles <-
  fhircrackr:::fhir_bundle_list(condition_bundles)

#TODO löschen
fhir_save(bundles = encounter_bundles, directory = "Bundles/Encounters")
fhir_save(bundles = condition_bundles, directory = "Bundles/Conditions")

enc_description <- fhir_table_description(
  "Encounter",
  cols = c(
    encounter.id = "id",
    subject = "subject/reference",
    encounter.start = "period/start",
    encounter.end = "period/end",
    diagnosis = "diagnosis/condition/reference",
    diagnosis.use.code = "diagnosis/use/coding/code",
    diagnosis.use.system = "diagnosis/use/coding/system",
    serviceType = "serviceType/coding/display"
  )
)

message("Cracking ", length(encounter_bundles), " Encounter Bundles.\n")

encounters <- fhir_crack(
  encounter_bundles,
  design = enc_description,
  brackets = brackets,
  sep = sep,
  data.table = TRUE,
  verbose = 0
)

rm(encounter_bundles)


con_description <- fhir_table_description(
  "Condition",
  cols = c(
    condition.id = "id",
    clinicalStatus.code = "clinicalStatus/coding/code",
    clinicalStatus.system = "clinicalStatus/coding/system",
    verificationStatus.code = "verificationStatus/coding/code",
    verificationStatus.system = "verificationStatus/coding/system",
    code = "code/coding/code",
    code.system = "code/coding/system",
    subject = "subject/reference",
    encounter = "encounter/reference"
  )
)

message("Cracking ", length(condition_bundles), " Condition Bundles.\n")

conditions <- fhir_crack(
  condition_bundles,
  design = con_description,
  brackets = brackets,
  sep = sep,
  data.table = TRUE,
  verbose = 0
)

rm(condition_bundles)


if (nrow(encounters) == 0) {
  write(
    "Konnte keine Encounter-Ressourcen zu den gefundenen Patients finden. Abfrage abgebrochen.",
    file = "errors/error_message.txt"
  )
  stop("No Encounters for Patients found - aborting.")
}

### generate conditions table --> has all conditions of all Patients in the initial population
if (nrow(conditions) > 0) {
  #extract diagnosis use info from encounter table
  useInfo <-
    fhir_melt(
      encounters,
      columns = c("diagnosis", "diagnosis.use.code", "diagnosis.use.system"),
      brackets = brackets,
      sep = sep,
      all_columns = TRUE
    )
  
  useInfo <- fhir_rm_indices(useInfo, brackets = brackets)
  
  useInfo <-
    useInfo[, c("encounter.id",
                "diagnosis",
                "diagnosis.use.code",
                "diagnosis.use.system")]
  
  useInfo[, diagnosis := sub("Condition/", "", diagnosis)]
  
  # expand condition codes
  for (i in 1:2) {
    conditions <-
      fhir_melt(
        conditions,
        columns = c("code", "code.system"),
        brackets = brackets,
        sep = sep,
        all_columns = TRUE
      )
  }
  
  
  conditions <- fhir_rm_indices(conditions, brackets = brackets)
  conditions[, resource_identifier := NULL]
  
  # filter for ICD codesystem
  conditions <- conditions[grepl("icd-10", code.system)]
  
  # add diagnosis use info to condition table
  message(
    "Merging Condition and Encounter data based on Condition id.\n",
    "Number of unique Condition ids in Condition data: ",
    length(unique(conditions$condition.id)),
    " in ",
    nrow(conditions),
    " rows",
    "\n",
    "Number of unique Condition ids in Encounter data: ",
    length(unique(useInfo$diagnosis)),
    " in ",
    nrow(useInfo),
    " rows",
    "\n"
  )
  
  conditions <- merge.data.table(
    x = conditions,
    y = useInfo,
    by.x = "condition.id",
    by.y = "diagnosis",
    all.x = TRUE
  )
  
  # prepare key variables
  conditions[, subject := sub("Patient/", "", subject)]
  conditions[, encounter := sub("Encounter/", "", encounter)]
  
  # merge encounter ids coming from the encounter.id vs. ids coming from the condition.encounter element into one column
  conditions[is.na(encounter.id), encounter.id := encounter]
  conditions[, encounter := NULL]
}


### prepare encounter table
# remove diagnosis info and indices
encounters[, c("diagnosis", "diagnosis.use.code", "diagnosis.use.system") :=
             NULL]

encounters <- fhir_rm_indices(encounters, brackets = brackets)

# prepare key variable for merge
encounters[, subject := sub("Patient/", "", subject)]

# sort out col types
encounters[, encounter.start := as.Date(encounter.start)]
encounters[, encounter.end := as.Date(encounter.end)]

# merge based on subject id and temporal relation of observation date and encounter times
message(
  "Merging Observation and Encounter data based on Subject id and time.\n",
  "Number of unique Subject ids in Observation data: ",
  length(unique(obsdata$subject)),
  " in ",
  nrow(obsdata),
  " rows",
  "\n",
  "Number of unique Subject ids in Encounter data: ",
  length(unique(encounters$subject)),
  " in ",
  nrow(encounters),
  " rows",
  "\n"
)
# encounters <- encounters[!grepl(".*-E-[0-9]-A", encounters$encounter.id)]
cohort <-
  obsdata[encounters, on = .(subject,
                             NTproBNP.date >= encounter.start,
                             NTproBNP.date <= encounter.end),
          c("encounter.id",
            "encounter.start",
            "encounter.end",
            "serviceType") := list(encounter.id, encounter.start, encounter.end, serviceType)][]

rm(obsdata)

# remove encounters that don't have a NTproBNP observation within their encounter.period
cohort <-
  cohort[NTproBNP.date >= encounter.start &
           NTproBNP.date <= encounter.end]

# filter conditions: only keep conditions belonging to the encounters we have just filtered
if (nrow(conditions) > 0) {
  conditions <- conditions[encounter.id %in% cohort$encounter.id]
}


###Export
#TODO schön machen
write.csv2(cohort, paste0("Ergebnisse/Kohorte.csv"))
write.csv2(conditions, paste0("Ergebnisse/Diagnosen.csv"))

# logging
runtime <- Sys.time() - start

con <- file("Ergebnisse/smith_select.log")
write(
  paste0(
    "smith_select.R finished at ",
    Sys.time(),
    ".\n",
    "Extracted ",
    length(cohort$encounter.id),
    " Encounters based on ",
    length(unique(cohort$subject)),
    " Patients.\n",
    "R script execution took ",
    round(runtime, 2),
    " ",
    attr(runtime, "units"),
    "."
  ),
  file = con
)
close(con)
